### PR TITLE
Rebuilding enumerable methods in ruby

### PR DESCRIPTION
**Pull request by Maha-Magdy**
What we did in this project, my partner and I is to re-create some enumerable methods and add them to the enumerable module. 
The methods that we created are:
#my_each
#my_each_with_index
#my_all?
#my_any?
#my_none?
#my_count
#my_map
#my_inject
, and also we made #my_multiply_els which uses my_inject to multiply all the elements from the array given.

### CODE REVIEW

**Review by umairarshadbutt — CHANGES_REQUESTED** [review]
# Status: REQUIRED CHANGES :warning: :construction: 

Hi @Maha-Magdy and @Anvi98 
Such a remarkable project you made :+1: ! Awesome README :100: !
We, however, do have a few changes to carry out before we proceed.

##

Kindly work on the suggested changes.
- [x] `#my_each` when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- [x] `#my_each` when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [x] `#my_each_with_index` when a block is given when self is an array returns the array itself after calling the given block once for each element in self
- [x] `#my_each_with_index` when a block is given when self is a range returns the range itself after calling the given block once for each element in self
- [x] `#my_all?` when no block or argument is given returns false when one of the collection members are false or nil
- [x] `#my_all?` when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
- [x] `#my_all?` when a class is passed as an argument returns true if all of the collection is a member of such class::Numeric
- [x] `#my_all?` when a class is passed as an argument returns false if any of the collection is not a member of such class
- [x] `#my_all?` when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] `#my_all?` when a Regex is passed as an argument returns false if any of the collection does not match the Regex
- [x] `#my_all?` when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [x] `#my_all?` when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
- [x] `#my_any?` when no block or argument is given returns false if at least one of the collection is not true
- [x] `#my_any?` when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
- [x] `#my_any?` when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
- [x] `#my_any?` when a Regex is passed as an argument returns true if any of the collection matches the Regex
- [x] `#my_any?` when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] `#my_any?` when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- [x] `#my_any?` when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern
- [x] `#my_none?` when no block or argument is given returns true only if none of the collection members is true
- [x] `#my_none?` when a class is passed as an argument returns true if none of the collection is a member of such class
- [x] `#my_none?` when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
- [x] `#my_none?` when a class is passed as an argument returns false if any of the collection is a member of such class
- [x] `#my_none?` when a Regex is passed as an argument returns true if none of the collection matches the Regex
- [x] `#my_none?` when a Regex is passed as an argument returns false if any of the collection matches the Regex
- [x] `#my_none?` when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [x] `#my_none?` when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
- [x] `#my_count` returns the number of items in enum through enumeration
- [x] `#my_count` returns the number of items in enum through enumeration::range
- [x] `#my_count` counts the number of elements yielding a true value if a block is given
- [x] `#my_count` does not mutate the original array
- [x] `#my_map` returns an Enumerator if no block is given
- [x] `#my_map` executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
- [x] `#my_inject` does not mutate the original array
- [x] `#my_inject` when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
- [x] `#my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
- [x] `#my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
- [x] `#my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
- [x] `#my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
##
- **When you have made the specified changes, please submit another code review. Thank you!**
##
**If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

**Comment by Anvi98** [comment]
In this new PR, we tried to fix all the observations from the last TSE.  However, we want to highlight something in our code.
we commented this code below in the my_all?, my_any? my_none? enumerable method:

# elsif !param.is_a?(Class) && !param.is_a?(Regexp)
        #   case param
        #     in self
        #       return true
        #   else
        #     return false
        #   end

The reason for that is because we tried to implement the pattern matching with the case statement, and this is in the experimental state. So, to pass the checks, we commented on this code, but if you need to test the full code, you need to uncomment this small part of it.

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,213 @@
...
+      end
+    else
+      to_a.my_each do |item|
+        return my_all_flag = false if item.nil? || item == false
+      end
+    end
+    my_all_flag
+  end
+
+  def my_any?(param = nil)
```
Comment (smcommits) [inline enumerable.rb:70]: - [ ] **enumerables #my_any? when no block or argument is given returns false if at least one of the collection is not true**

Case
```
false_array.my_any?
```
```
expected false
            got true
```

Comment (smcommits) [inline enumerable.rb:66]: - [x] **enumerables #my_any? when no block or argument is given returns false if at least one of the collection is not true**

Case
```
false_array.my_any?
```
```
expected false
            got true
```

- [x] **enumerables #my_any? when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern**

Case
```
words.my_any?('cat')
```
```
expected true
            got false
```

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,213 @@
...
+          return my_any_flag if item.nil? || item == false
+        end
+        my_any_flag = false
+      end
+    else
+      to_enum(:my_map)
+    end
+  end
+
+  def my_inject(param = nil, symbol_value = nil)
```
Comment (smcommits) [inline enumerable.rb:170]: - [x] **enumerables #my_inject raises a "LocalJumpError" when no block or argument is given**

Case
```
array.my_inject
```
```
expected LocalJumpError but nothing was raised
```

**Review by smcommits — CHANGES_REQUESTED** [review]
# Status: Changes Requested 
## Hello, @Maha-Magdy and @Anvi98! You guys have done a great job. The methods are passing almost all the test cases. But, your methods are still missing some behaviors that are present in the original methods

## Request Changes
**Please note that I have uncommented the code that you suggested and the following cases are still failing**

I have used the following variables for testing the methods. The inline code shows all the use cases of these variables and their respective output. 

```
words = %w[dog door rod blade cat]
false_array = [nil, false, nil, false]
```

#
### Please don't hesitate to contact me on [Slack](https://microverse-students.slack.com/team/U017N7LLZ40) if you need assistance or believe I have overlooked something and made an unfair request.

**Review by rukundoeric — CHANGES_REQUESTED** [review]
# Status: ♻️ CHANGES REQUIRED ♻️
_Great job guys so far 👏 💯 👍. However, there are few issues we need to address before we move on, please check the comments below._

- ### To Highlight
  - [x] The project is clean. 💯
  - [x] The project follows Git flow. 💯

- ### To Improve 
  - [x] [MANDATORY] #my_all? when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern.
     - Example `[3, 3, 3].my_all?(3)`
       - Expected: `true`
       - Got: `nil`
  - [x] [MANDATORY] #my_any? when no block or argument is given returns true if at least one of the collection is not false or nil
     - Example `[3, 3, 4].my_any?`
       - Expected: `true`
       - Got: `false`
  - [x] [MANDATORY] #my_none? when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern
     - Example `[5, 'nino', 8 , 9].my_none?(5)`
       - Expected: `false`
       - Got: `nil`
  - [x] [MANDATORY] #my_none? when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
     - Example `[2, 'nino', 8 , 9].my_none?(5)`
       - Expected: `true`
       - Got: `nil`
  - [x] [MANDATORY] #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
     - Example `[10, 11, 2, 1].my_map(&proc { |num| num > 10 }) { |num| num < 10 }`
       - Expected: `[false, true, false, false]`
       - Got: `[true, false, true, true]`
  - [x] [MANDATORY] Kindly remove all commented codes.     

_Please make all the required changes before you submit another code review. If you have any problems or concerns, you can reach out to me on Slack: Rukundo Eric 🚀 . I wish you all the very best 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/), 
Slack: Rukundo Eric**

**Review by rukundoeric — CHANGES_REQUESTED** [review]
# Status: ♻️ CHANGES REQUIRED ♻️
_Great job guys so far 👏 💯 👍. However, there are few issues we need to address before we move on, please check the comments below._

- ### To Highlight
  - [x] The project is clean. 💯
  - [x] The project follows Git flow. 💯

- ### To Improve  
  - [ ] [MANDATORY] #my_map executes only the proc when both a block and a proc are given and returns a new array with the results of running the proc once for every element in enum.
     - Example `[10, 11, 2, 1].my_map(&proc { |num| num > 10 }) { |num| num < 10 }`
       - Expected: `[false, true, false, false]`
       - Got: `both block arg and actual block given`.

  NOTE: Remember if both proc and block are given. execute proc and ignore block.

_Please make all the required changes before you submit another code review. If you have any problems or concerns, you can reach out to me on Slack: Rukundo Eric 🚀 . I wish you all the very best 🥇_

### _Keep working hard._
### _Happy Coding😃_
-----------------------------------------------------------------------------------------------------------------------------------------
**Regards,
[Rukundo Eric](https://rukundo-eric.herokuapp.com/) ,
[LinkedIn](https://www.linkedin.com/in/rukundo-eric/), 
Slack: Rukundo Eric**

**Review by henatan99 — APPROVED** [review]
# Project Approved 
## Congratulations :100: :partying_face: 
Hey @Maha-Magdy, you did an amazing job. Your enumerable module passes all the spec tests. Well Done!!!!! 
You may now merge the PR and submit the mandatory project completion form. 

If you have any queries, you can reach out to me via:
- [LinkedIn](https://www.linkedin.com/in/henok-mekonnen-2a251613/)
- [Twitter](https://twitter.com/henatan99)
- or Slack by 'Henok Mekonnen Mossissa' 

### Good Luck in your upcoming projects

**Pull request merged**